notmuch: install the vim plugin

RyanGibb opened the pull request:
Seems I messed up squashing https://github.com/NixOS/nixpkgs/pull/308318

> Changes look good, but I haven't tested them. In any case, could you squash the commits please into 1 notmuch: install the vim plugin?

That's fixed now @doronbehar.

I'm still thinking about how to pick up the Ruby bindings. It would make it easier if they were in the same output. It doesn't look like there were any strong reasons against this https://github.com/NixOS/nixpkgs/pull/155056.

Julow committed:
notmuch: install the vim plugin

Install the vim plugin to $out/share/vim-plugins/notmuch and
$out/share/nvim/site/plugin/notmuch.vim. It is only static files and
doesn't add any dependency so it is enabled by default. Add `withVim`
argument to override this.

Co-authored-by: Jules Aguillon <juloo.dsi@gmail.com>

RyanGibb committed:
notmuch: vim ruby bindings and gems

We removed the ruby output so we will be able to reference the ruby files from
the vim plugin, which is in $out. If we try to reference $ruby from $out we get
a cyclic dependency.

RyanGibb commented:
Okay, I've moved the Ruby bindings to `$out` along with the vim plugins. I've also added `rubyPackages.mail` to `RUBYLIBS`. This works!

There's two issues I'm having with separate outputs: I'm not picking up vim plugins from the separate vim output for whatever reason, but more importantly Nix complains about cyclic dependencies if we reference the ruby output from the vim output, so we'd require duplicating them.

doronbehar commented on pkgs/applications/networking/mailreaders/notmuch/default.nix:
```diff
@@ -133,6 +134,14 @@ stdenv.mkDerivation rec {
+    ln -s $out/share/vim-plugins/notmuch $out/share/nvim/site
+  '' + lib.optionalString (withVim && withRuby) ''
+    cat >> $out/share/vim-plugins/notmuch/plugin/notmuch.vim << EOF
+      let \$RUBYLIB=\$RUBYLIB . ":$out/${ruby.libPath}/${ruby.system}:${rubyPackages.mail}/lib/ruby/gems/${ruby.version.libDir}/gems/mail-${rubyPackages.mail.version}/lib"
```
Hmm I'm surprised there is no shortcut for:

```nix
${rubyPackages.mail}/lib/ruby/gems/${ruby.version.libDir}/gems/mail-${rubyPackages.mail.version}/lib"
```

The python equivalent would have been simply:

```nix
${pythonPackages.mail}/${pythonPackages.python.sitePackages}
```

It would have been nice if `ruby.gems` had provided such an equivalent, but according to my poor `git grep` searches it doesn't seem like a common thing to do - to set `RUBYLIB` to such a search path of both ruby and a few gems.

Either way, it'd be nice to be able to define this variable a bit more abstractly, and generally in the `passthru` like this (untested):

```nix
passthru = {
  # ...
  # Perhaps this should be part of the `ruby.passthru`, or an attribute
  # for all `ruby.gems`.
  getGemRubyLibPath = gem: ${gem.ruby.gemPath}/gems/${gem.gemName}-${gem.version}/lib";
  rubyLibPath = lib.concatStringsSep ":" [
    "$out/${ruby.libPath}/${ruby.system}"
  ] ++ lib.map finalAttrs.passthru.getGemRubyLibPath [
    ruby.gems.mail
  ];
};
```

Note how `getGemRubyLibPath` does what you wanted to do much more elegantly. With the above, you can access the `RUBYLIB` path you want with:

```suggestion
      let \$RUBYLIB=\$RUBYLIB . ":${finalAttrs.passthru.rubyLibPath}"
```

You'd also have to transition to `finalAttrs` (like [this commit for instance](https://github.com/NixOS/nixpkgs/commit/c1f5835e6ba8b2df709d51fa0e81b1f51fd7105c)) in a separate commit, before this commit.

RyanGibb replied:
Thanks for the suggestions!

> You'd also have to transition to finalAttrs (like [this commit for instance](https://github.com/NixOS/nixpkgs/commit/c1f5835e6ba8b2df709d51fa0e81b1f51fd7105c)) in a separate commit, before this commit.

Can I ask why does this need to be in a separate commit?

RyanGibb replied:
Okay, I've changed the approach a bit to use `buildEnv`.

doronbehar commented:
> There's two issues I'm having with separate outputs: I'm not picking up vim plugins from the separate vim output for whatever reason, but more importantly Nix complains about cyclic dependencies if we reference the ruby output from the vim output, so we'd require duplicating them.

I'm not sure I understood this, but I trust you you made an effort and reached the correct conclusion. Removing the `ruby` output is definitely shouldn't be a big deal.

RyanGibb commented:
> I'm not sure I understood this, but I trust you you made an effort and reached the correct conclusion. Removing the ruby output is definitely shouldn't be a big deal.

Re-reading this I don't think I made much sense actually. We wouldn't *require* deduplicating them, we'd just be duplicating the ruby bindings across both outputs. But yeah, that sounds good.

doronbehar commented:
> Can I ask why does this need to be in a separate commit?

IMO it makes the changes to the expression laid out better, and easier to review, both in the far and the near future. It really is a change independent from the other changes.

RyanGibb commented:
It turns out there was issues with the mail gem, as we weren't setting RUBYLIB for it's dependencies. I've squash pushed a fix adapted from https://github.com/arcnmx/nixexprs/blob/master/pkgs/vimPlugins.nix that seems to work.

RyanGibb commented on pkgs/applications/networking/mailreaders/notmuch/default.nix:
```diff
@@ -6,12 +6,14 @@
 , pythonPackages
 , emacs
 , ruby
+, rubyPackages
```
Done!

doronbehar commented on pkgs/applications/networking/mailreaders/notmuch/default.nix:
```diff
@@ -133,7 +133,28 @@ stdenv.mkDerivation rec {
+    mkdir -p $out/share/nvim
+    ln -s $out/share/vim-plugins/notmuch $out/share/nvim/site
+  '' + lib.optionalString (withVim && withRuby) (let
+    gemEnv = buildEnv {
```
Interesting! Now I'm surprised how `ruby.withPackages` doesn't fit here, because it doesn't add `"/lib"` to `pathsToLink` (found this out locally).

In anycase, it'd still be great if you'd transition into `finalAttrs`, as it would allow someone to override this derivation if they need from some reason more gems installed.

Here are changes I tried, based upon yours:

```diff
diff --git i/pkgs/applications/networking/mailreaders/notmuch/default.nix w/pkgs/applications/networking/mailreaders/notmuch/default.nix
index c9b0e158acf2..8348919d1769 100644
--- i/pkgs/applications/networking/mailreaders/notmuch/default.nix
+++ w/pkgs/applications/networking/mailreaders/notmuch/default.nix
@@ -15,12 +15,12 @@
 , withVim ? true
 }:
 
-stdenv.mkDerivation rec {
+stdenv.mkDerivation (finalAttrs: {
   pname = "notmuch";
   version = "0.38.3";
 
   src = fetchurl {
-    url = "https://notmuchmail.org/releases/notmuch-${version}.tar.xz";
+    url = "https://notmuchmail.org/releases/notmuch-${finalAttrs.version}.tar.xz";
     hash = "sha256-mvRsyA2li0MByiuu/MJaQNES0DFVB+YywPPw8IMo0FQ=";
   };
 
@@ -137,29 +137,27 @@ stdenv.mkDerivation rec {
     make -C vim DESTDIR="$out/share/vim-plugins/notmuch" prefix="" install
     mkdir -p $out/share/nvim
     ln -s $out/share/vim-plugins/notmuch $out/share/nvim/site
-  '' + lib.optionalString (withVim && withRuby) (let
-    gemEnv = buildEnv {
-      name = "notmuch-vim-gems";
-      paths = with ruby.gems; [ mail ];
-      pathsToLink = [ "/lib" "/nix-support" ];
-    };
-  in ''
+  '' + lib.optionalString (withVim && withRuby) ''
     PLUG=$out/share/vim-plugins/notmuch/plugin/notmuch.vim
     cat >> $PLUG << EOF
-      let \$GEM_PATH=\$GEM_PATH . ":${gemEnv}/${ruby.gemPath}"
+      let \$GEM_PATH=\$GEM_PATH . ":${finalAttrs.passthru.gemEnv}/${ruby.gemPath}"
       let \$RUBYLIB=\$RUBYLIB . ":$out/${ruby.libPath}/${ruby.system}"
       if has('nvim')
     EOF
-    for gem in ${gemEnv}/${ruby.gemPath}/gems/*/lib; do
+    for gem in ${finalAttrs.passthru.gemEnv}/${ruby.gemPath}/gems/*/lib; do
       echo "ruby \$LOAD_PATH.unshift('$gem')" >> $PLUG
     done
     echo 'endif' >> $PLUG
-  '');
+  '';
 
   passthru = {
-    pythonSourceRoot = "notmuch-${version}/bindings/python";
+    pythonSourceRoot = "notmuch-${finalAttrs.version}/bindings/python";
     tests.version = testers.testVersion { package = notmuch; };
-    inherit version;
+    gemEnv = buildEnv {
+      name = "notmuch-vim-gems";
+      paths = with ruby.gems; [ mail ];
+      pathsToLink = [ "/lib" "/nix-support" ];
+    };
 
     updateScript = gitUpdater {
       url = "https://git.notmuchmail.org/git/notmuch";
@@ -176,4 +174,4 @@ stdenv.mkDerivation rec {
     platforms   = platforms.unix;
     mainProgram = "notmuch";
   };
-}
+})
```

And here is how I tested the overriding mechanism:

Without overriding:

```
$ nix show-derivation --impure --expr 'with (builtins.getFlake "'$PWD'").legacyPackages.${builtins.currentSystem}; notmuch.overrideAttrs(old: {
})' | jq --raw-output '.[].env.postInstall'
moveToOutput bin/notmuch-emacs-mua $emacs
make -C bindings/ruby install \
  vendordir=$out/lib/ruby \
  SHELL=$SHELL \
  $makeFlags "${makeFlagsArray[@]}" \
  $installFlags "${installFlagsArray[@]}"
cp notmuch-git $out/bin/notmuch-git
wrapProgram $out/bin/notmuch-git --prefix PATH : $out/bin:/nix/store/apl2ibxqrxbbj6z79nfgv892999z1zyv-git-2.44.0/bin
make -C vim DESTDIR="$out/share/vim-plugins/notmuch" prefix="" install
mkdir -p $out/share/nvim
ln -s $out/share/vim-plugins/notmuch $out/share/nvim/site
PLUG=$out/share/vim-plugins/notmuch/plugin/notmuch.vim
cat >> $PLUG << EOF
  let \$GEM_PATH=\$GEM_PATH . ":/nix/store/a4s4ymn7k90yh9bld5zic587n1kg0k1v-notmuch-vim-gems/lib/ruby/gems/3.1.0"
  let \$RUBYLIB=\$RUBYLIB . ":$out/lib/ruby/3.1.0/x86_64-linux"
  if has('nvim')
EOF
for gem in /nix/store/a4s4ymn7k90yh9bld5zic587n1kg0k1v-notmuch-vim-gems/lib/ruby/gems/3.1.0/gems/*/lib; do
  echo "ruby \$LOAD_PATH.unshift('$gem')" >> $PLUG
done
echo 'endif' >> $PLUG
```

With overriding:

```
$ nix show-derivation --impure --expr 'with (builtins.getFlake "'$PWD'").legacyPackages.${builtins.currentSystem}; notmuch.overrideAttrs(old: {
  passthru = old.passthru // {
    gemEnv = buildEnv {
      name = "notmuch-more-gems";
      paths = with ruby.gems; [
        mail
        clamp
      ];
    };
  };
})' | jq --raw-output '.[].env.postInstall'
moveToOutput bin/notmuch-emacs-mua $emacs
make -C bindings/ruby install \
  vendordir=$out/lib/ruby \
  SHELL=$SHELL \
  $makeFlags "${makeFlagsArray[@]}" \
  $installFlags "${installFlagsArray[@]}"
cp notmuch-git $out/bin/notmuch-git
wrapProgram $out/bin/notmuch-git --prefix PATH : $out/bin:/nix/store/apl2ibxqrxbbj6z79nfgv892999z1zyv-git-2.44.0/bin
make -C vim DESTDIR="$out/share/vim-plugins/notmuch" prefix="" install
mkdir -p $out/share/nvim
ln -s $out/share/vim-plugins/notmuch $out/share/nvim/site
PLUG=$out/share/vim-plugins/notmuch/plugin/notmuch.vim
cat >> $PLUG << EOF
  let \$GEM_PATH=\$GEM_PATH . ":/nix/store/hadibs48n5cympz47k3ypag89yvbd6x1-notmuch-more-gems/lib/ruby/gems/3.1.0"
  let \$RUBYLIB=\$RUBYLIB . ":$out/lib/ruby/3.1.0/x86_64-linux"
  if has('nvim')
EOF
for gem in /nix/store/hadibs48n5cympz47k3ypag89yvbd6x1-notmuch-more-gems/lib/ruby/gems/3.1.0/gems/*/lib; do
  echo "ruby \$LOAD_PATH.unshift('$gem')" >> $PLUG
done
echo 'endif' >> $PLUG
```

Note how the path to the `gemEnv` changed - as requested. It won't be possible to perform such a change so elegantly without `finalAttrs`.

RyanGibb replied:
Okay, makes sense!

doronbehar replied:
But @RyanGibb you didn't define `gemEnv` in `passthru`, and then used `finalAttrs.passthru`...

doronbehar reviewed (changes requested):
P.S it would be nice to explain in the commit message why the `ruby` output was discarded.

RyanGibb committed:
notmuch: use finalAttrs

Co-authored-by: Doron Behar <doron.behar@gmail.com>

RyanGibb commented:
I've transitioned to finalAttrs and added a commit message to explain why the ruby output was discarded.

doronbehar reviewed (changes requested):
@RyanGibb it looks better overall, but it doesn't make sense to put the `finalAttrs` commit at the end. It makes much more sense to put it before the `withRuby && withVim` commit, as it makes the overall complexity of the diffs much cleaner.

Also, the explanation in the commit message about the removal of the `ruby` output is not phrased well. I'd write it as follows (format it of course to Git commit messages' 80 characters format:

> We removed the ruby output so we will be able to reference the ruby files from the vim plugin, which is in `$out`. If we try to reference $ruby from $out we get a cyclic dependency.

No need to mention the idea of duplicating `$ruby` and `$out`, if you don't explain why you don't do that :).

doronbehar commented:
BTW, how can one test this new feature? What exactly I need to run in Vim / Neovim?

doronbehar commented:
Also, naturally if the `finalAttrs` commit will be 2nd, it would simply be:

```
notmuch: use finalAttrs
```

RyanGibb commented:
> @RyanGibb it looks better overall, but it doesn't make sense to put the `finalAttrs` commit at the end. It makes much more sense to put it before the `withRuby && withVim` commit, as it makes the overall complexity of the diffs much cleaner.

Sure

> No need to mention the idea of duplicating $ruby and $out, if you don't explain why you don't do that :).

But I did explain that we don't do that to avoid duplicating the bindings? I'm not too bothered about this though

> Also, naturally if the finalAttrs commit will be 2nd, it would simply be:

I've rebased and changed the order

> BTW, how can one test this new feature? What exactly I need to run in Vim / Neovim?

Try `vim -c NotMuch`. It might not show anything much without a notmuch index though.

RyanGibb commented:
Agree! Thanks for the suggestion :-)

doronbehar commented:
> Agreed! Thanks for the suggestion :-)

Great :) (I'm waiting for you to `git push`).

RyanGibb commented:
Whoops, looks like I pushed the wrong commit before! I've pushed the right one now

doronbehar reviewed (approved):
Works great!

doronbehar commented:
From some reason I didn't get a notification for the last force push, sorry it took a while :). Thanks for your cooperation and for insisting on contributing this!

RyanGibb commented:
No worries! Thanks for the help reviewing @doronbehar :-)
